bylaws/008-psr-evolution.md

```diff
@@ -0,0 +1,69 @@
+
+This document defines a process to be followed in updating PSR interfaces, in a way that is not breaking in regard to behavior for end users, and with an appropriate upgrade path for the packages.
+
+## New releases
+
+A new minor release of a PHP-FIG package containing interfaces for a PSR MUST follow these rules:
+ * the new release MUST follow [Semantic Versioning](https://semver.org/) rules;
+ * the PSR behavior MUST NOT be altered;
+ * the packages that implement the interfaces or users that are consuming them MUST NOT suffer breaking changes;
+ * the PHP version constraint of the PHP-FIG package MAY be altered to require newer language features that would aid cross-compatibility;
+ * the PHP version constraint of the PHP-FIG package MUST NOT be altered to use newer language features that would create cross-compatibility issues.
+ 
+A new major release of a PHP-FIG package containing interfaces for a PSR MUST follow the same rules, with this exception:
+ * the new major version of the package MAY contain breaking changes if the implementing packages have a reasonable upgrade path, like the possibility of releasing a cross-compatible implementation with the previous releases;
+ * the new major version of the package MAY refer to a new, superseding PSR.
+
+### Workflow
+
+Since releasing new versions of the interfaces MUST NOT alter the PSR in its behavior, those releases can be voted in with the same process as errata changes. The new releases MUST be declared and embedded in the PSR document, with eventual indications on the upgrade path in the meta document.
+
+### Practical example
+
+A common case for an upgrade in the interfaces is to add types for parameters and returns, since they are new language features introduced by PHP 7, and many PSR interfaces predate that release. We'll use a method from PSR-11 as an example of how a PSR interface could be updated.
+
+#### PSR-11: the interface
+
+PSR-11 is released with the [`psr/container` package](https://packagist.org/packages/psr/container) and it holds the `ContainerInterface`, which has this method:
+```php
+    /**
+     * @param string $id Identifier of the entry to look for.
+     *
+     * @return bool
+     */
+    public function has($id);
+```
+This method could be updated with a new minor release that adds the argument type for `$id`:
+```php
+public function has(string $id);
+```
+This change would technically be a breaking change, but thanks to the [limited contravariance possible in PHP 7.2](https://wiki.php.net/rfc/parameter-no-type-variance), we can avoid that. This means that just by requiring `"php": "^7.2"` in the `psr/container` `composer.json`, we could tag this change as a minor release, and have all the implementors be automatically cross-compatible, provided that they declare `"psr/container": "^1.0"` (or equivalent) as a constraint, which they should.
```

Comment:
I thought we were going to go with major versions here to avoid the different-environment issue that @nicolas-grekas pointed out.

Reply:
This could be discussed. I understand Nicolas motivation, but OTOH this approach would not require action from the implementing libraries on the first step, and that's why I preferred it.

Reply:
I've thought twice about this: given that no apps require any psr packages directly, it means that they come transitively. If we go with v2 here, and if all libs update their composer.json to allow it, then the different-environment issue appears also.
This means my argument is fragilized here, maybe not worth it. For thoughts.

Reply:
Actually, I believe that the expectancy would be that apps that calls the interfaces in their code (and only link interfaces to implementation some other code) actually require psr packages directly.
As this is what they should do. If you use a namespace in your code, you should have the package providing it in composer.json.

But I also believe that apps do not require psr packages. And a minor change will trigger a composer update to upgrade it to 1.1 from 1.0.
And mainly that's what I also argued earlier that a major version is needed as otherwise $container->has(1) that returns false, after change, will throw a `TypeError`.

Reply:
> [...] otherwise $container->has(1) that returns false, after change, will throw a TypeError.

That's true only if you're using `strict_types=1`, and only at the implementation change, not on the interface change; so the implementation can choose on its own if the change warrants a major or not.

Reply:
Thank you @Jean85 for explaining it again. You're right, I understood it now. It's true that IDE might complain about you using an interface method with a wrong parameter type but at run-time the check is done against the actual implementing class, not against the interface... obviously, I would say now.

I don't have any other issue with it being a minor version of psr package with php ^7.2 required.